samza-log4j2/src/main/java/org/apache/samza/logging/log4j2/StreamAppender.java

```diff
@@ -75,16 +75,16 @@
 
   private static final String JAVA_OPTS_CONTAINER_NAME = "samza.container.name";
   private static final String JOB_COORDINATOR_TAG = "samza-job-coordinator";
-  private static final String SOURCE = "log4j-log";
+  protected static final String SOURCE = "log4j-log";
 
   // Hidden config for now. Will move to appropriate Config class when ready to.
   private static final String CREATE_STREAM_ENABLED = "task.log4j.create.stream.enabled";
 
   private static final long DEFAULT_QUEUE_TIMEOUT_S = 2; // Abitrary choice
   private final BlockingQueue<byte[]> logQueue = new LinkedBlockingQueue<>(DEFAULT_QUEUE_SIZE);
 
-  private SystemStream systemStream = null;
-  private SystemProducer systemProducer = null;
+  protected SystemStream systemStream = null;
+  protected SystemProducer systemProducer = null;
```

Comment:
Just to reiterate my comment on the other [PR](https://github.com/apache/samza/pull/1470) for protected changes. I see multiple attributes being exposed here where in reality you probably only need few attributes like source, key that needs to be overridden but still keeping the logic of sending the event to underlying producer same.

The way I see it is not about one liner methods rather conceptual overrides that extenders need to provide. This way it still avoids code duplication, unnecessary exposure (against closed principle) and also adds burden to reason about all the exposed attributes.

e.g. what state should this attributes be before sending event and potential differences in the implementors perception into those contracts.

TLDR; I feel you should have overrides for what you need as opposed to exposing all that is required to perform a functionality if the functionality is common enough across the hierarchy.

Reply:
thanks for the detailed explanation.
the method that was exposed in previous pr #1470 is 
`protected void sendEventToSystemProducer(byte[] serializedLogEvent) {
       metrics.logMessagesBytesSent.inc(serializedLogEvent.length);
       metrics.logMessagesCountSent.inc();
      systemProducer.send(SOURCE, new OutgoingMessageEnvelope(systemStream, keyBytes, serializedLogEvent));
 }`

the  functionality  that needs to be overriden is 
`systemProducer.send(SOURCE, new OutgoingMessageEnvelope(systemStream, keyBytes, serializedLogEvent));`

to be able to override this to have OutgoingMessageEnvelope have no partition key and no record key, the overrding statement will look like 

`systemProducer.send(SOURCE, new OutgoingMessageEnvelope(systemStream, null, null, serializedLogEvent));`

and to achieve this all 3 attributes systemproducer, source and systemstream are needed.

Please let me know if there is a  better way to achieve no partition and no record key in the OME.

Reply:
It looks like you need some sort of decorator around the actual log events that gets sent to the underlying producer. The decorator will take in a `serializedLogEvent` as `byte[]` and returns an `OutgoingMessageEnvelope` which the extenders can choose to override however they want.

For e.g

```
// default implementation
protected OutgoingMessageEnvelope decorateLogEvent(byte[] serializedLogEvent) {
   return new OutgoingMessageEnvelope(systemStream, keyBytes, serializedLogEvent);
}

private void sendEventToSystemProducer(byte[] serializedLogEvent {
  metrics.logMessagesBytesSent.inc(serializedLogEvent.length);
  metrics.logMessagesCountSent.inc();
  systemProducer.send(SOURCE, decorateLogEvent(serializedLogEvent));
}
```

Reply:
thank you. thats a great suggestion. updated.